pyquil/api/_base_connection.py

```diff
@@ -260,6 +261,217 @@ def qvm_run_payload(quil_program, classical_addresses, trials,
+    PyquilConfig, which in turn writes the refreshed auth credential to file.
+
+    Encapsulates the operations required for authorization & encryption
+      with the QPU.
+
+    Two operations are involved in authorization:
+      * Requesting & storing a user authentication token, used to authenticate calls
+        to Forest, Dispatch, and other Rigetti services
+      * Requesting a Curve ZeroMQ keypair for connection to the QPU. The response to
+        this request also comes with service endpoints: compiler server and QPU
+
+    The authentication tokens are of the standard JWT format and are issued by Forest Server.
+      The refresh token is only used to renew the access token, which is used for all transactions
+      and is valid for a short period of time.
+
+    In wrapping the PyQuilConfig object, it provides that object with a callback to
+      retrieve a valid engagement when needed, because the engagement is maintained here
+      but is used by the config to provide service endpoints.
+    """
+    def __init__(self, config: Optional[PyquilConfig] = None, lattice_name: Optional[str] = None, **kwargs):
+        super().__init__(**kwargs)
+        if config is not None:
+            self.config = config
+        else:
+            self.config = PyquilConfig()
+        self.config.get_engagement = self.engagement
+        self._engagement = None
+        self.headers.update(self.config.qcs_auth_headers)
+        self.headers['User-Agent'] = f"PyQuil/{__version__}"
+        self.lattice_name = lattice_name
+
+    def engage(self) -> None:
+        """
+        The heart of the QPU authorization process, `engage` makes a request to
+          the dispatch server for the information needed to communicate with the QPU.
+
+        This is a standard GraphQL request, authenticated using the access token
+          retrieved from Forest Server.
+
+        The response includes the endpoints to the QPU and QPU Compiler Server,
+          along with the set of keys necessary to connect to the QPU and the time at
+          which that key set expires.
+        """
+        query = '''
+          mutation Engage($name: String!) {
+            engage(input: { lattice: { name: $name }}) {
+              success
+              message
+              engagement {
+                type
+                qpu {
+                    endpoint
+                    credentials {
+                        clientPublic
+                        clientSecret
+                        serverPublic
+                    }
+                }
+                compiler {
+                    endpoint
+                }
+                expiresAt
+              }
+            }
+          }
+        '''
+        if not self.lattice_name:
+            raise ValueError("ForestSession requires lattice_name in order to engage")
+        logger.info(f"Requesting engagement from {self.config.dispatch_url}")
```

Comment:
I chose info to explain to the user, “what am I waiting on right now?”, since there is a few-second pause on script execution for calls to forest server and dispatch. From my use of it over the past few weeks, I’ve appreciated that, but if you want debug we’ll do debug

Reply:
fine with me